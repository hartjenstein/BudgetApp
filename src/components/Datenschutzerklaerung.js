import React from 'react';
import Header from './Header';

const Datenschutzerklaerung = () => (
  <div>
  <Header />
    <div className="content-container content">
      <h1>Datenschutzerklärung:</h1>


      <h2>Verantwortliche Stelle</h2>
      
      <p>Verantwortliche Stelle für die Erhebung, Verarbeitung und Nutzung Ihrer personenbezogenen Daten im Sinne des Bundesdatenschutzgesetzes (BDSG) ist
      
      N Hartjenstein, Richardstr. 101b, 12043 Berlin
      
      Sofern Sie der Erhebung, Verarbeitung oder Nutzung Ihrer Daten durch N Hartjenstein gemäß den datenschutzrechtlichen Bestimmungen insgesamt oder für einzelne Maßnahmen widersprechen möchten, richten Sie Ihren Widerspruch per Brief an folgende Kontaktdaten:
      
      Richardstr. 101b, 12043 Berlin
      
      Wir legen größten Wert auf den Schutz Ihrer Daten und die Wahrung Ihrer Privatsphäre. Nachstehend informieren wir Sie deshalb über die Erhebung und Verwendung persönlicher Daten bei Nutzung unserer Webseite.
      </p>
      
      <h2>Erheben, Speichern und Verwenden personenbezogener Daten</h2> 
        
      <p>Bei dem Besuch unserer Website erheben, verwenden und speichern wir Ihre personenbezogenen Daten. Personenbezogene Daten sind nach § 3 Abs. 1 BDSG Einzelangaben über persönliche oder sachliche Verhältnisse einer bestimmten oder bestimmbaren natürlichen Person (Betroffener).
        
        
        Datenerhebung unter strenger Beachtung der Gesetze
        
        Bei der Erhebung, dem Verwenden und Speichern von personenbezogenen Daten beachten wir die Vorschriften des Bundesdatenschutzgesetzes, des Telemediengesetzes und ggf. weiterer anwendbarer Datenschutzbestimmungen. Alle Kundendaten werden unter Beachtung der einschlägigen Vorschriften des Bundesdatenschutzgesetzes (BDSG) und des Telemediengesetzes (TMG) von uns gespeichert und verarbeitet.
        </p>
        <h2>Cookies</h2>
        
        <p>Unsere Internetseite verwendet an mehreren Stellen so genannte Cookies. „Cookies“ sind kleine Textdateien, die auf Ihrem Rechner abgelegt werden und die Ihr Browser speichert. Sie dienen dazu, unser Angebot nutzerfreundlicher, effektiver und sicherer zu machen. Des Weiteren ermöglichen Cookies unseren Systemen, Ihren Browser zu erkennen und Ihnen Services anzubieten. Cookies enthalten keine Personenbezogenen Daten. Sie haben die Möglichkeit, das Abspeichern von Cookies auf Ihrem Rechner durch entsprechende Einstellungen in Ihrem Browser zu verhindern. Hierdurch könnte allerdings der Funktionsumfang unseres Angebotes eingeschränkt werden.
        </p>
        
        <h2>Google-Analytics</h2>
        
        <p>Diese Website benutzt Google Analytics, einen Webanalysedienst der Google Inc. („Google“). Google Analytics verwendet sog. „Cookies“, Textdateien, die auf Ihrem Computer gespeichert werden und die eine Analyse der Benutzung der Website durch Sie ermöglichen. Die durch den Cookie erzeugten Informationen über Ihre Benutzung dieser Website werden in der Regel an einen Server von Google in den USA übertragen und dort gespeichert.
        Im Falle der Aktivierung der IP-Anonymisierung auf dieser Webseite wird Ihre IP-Adresse von Google jedoch innerhalb von Mitgliedstaaten der Europäischen Union oder in anderen Vertragsstaaten des Abkommens über den Europäischen Wirtschaftsraum zuvor gekürzt. Nur in Ausnahmefällen wird die volle IP-Adresse an einen Server von Google in den USA übertragen und dort gekürzt. Die IP-Anonymisierung ist auf dieser Website aktiv. Im Auftrag des Betreibers dieser Website wird Google diese Informationen benutzen, um Ihre Nutzung der Website auszuwerten, um Reports über die Websiteaktivitäten zusammenzustellen und um weitere mit der Websitenutzung und der Internetnutzung verbundene Dienstleistungen gegenüber dem Websitebetreiber zu erbringen.
        Die im Rahmen von Google Analytics von Ihrem Browser übermittelte IP-Adresse wird nicht mit anderen Daten von Google zusammengeführt. Sie können die Speicherung der Cookies durch eine entsprechende Einstellung Ihrer Browser-Software verhindern; wir weisen Sie jedoch darauf hin, dass Sie in diesem Fall gegebenenfalls nicht sämtliche Funktionen dieser Website vollumfänglich werden nutzen können. Sie können darüber hinaus die Erfassung der durch das Cookie erzeugten und auf Ihre Nutzung der Website bezogenen Daten (inkl. Ihrer IP-Adresse) an Google sowie die Verarbeitung dieser Daten durch Google verhindern, indem Sie das unter dem folgenden Link verfügbare Browser-Plugin herunterladen und installieren: http://tools.google.com/dlpage/gaoptout?hl=de
        </p>
        
        
        <h2>Auskunft</h2>

        <p>
        Sie haben jederzeit das Recht auf unentgeltliche Auskunft über Ihre gespeicherten Daten sowie das Recht auf Berichtigung, Löschung bzw. Sperrung. Wenden Sie sich hierzu bitte an die im Impressum angegebene Kontaktadresse. Wir stehen Ihnen jederzeit gern für weitergehende Fragen zu unseren Hinweisen zum Datenschutz und zur Verarbeitung Ihrer persönlichen Daten zur Verfügung.
        

        </p>
    </div>
    </div>
);

export default Datenschutzerklaerung;