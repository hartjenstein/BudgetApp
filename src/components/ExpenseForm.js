import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


//const date = new Date();
//moment example
const now = moment();
console.log(now.format('MMM Do, YYYY'));

// * using local state to keep track of the changes the user makes in the form
// only if user submits we send it of to redux
// * why did we switch back to the constructor function? We want to 
//populate the state with exitsing state which we get passed down from the edit expense page 
// via props. If we dont set up the contructor function there is no way to access the props! 
export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createAt) : moment(),
            calenderFocused: false,
            error: ''
        };
    };

    onDescriptionChange = e => {
        const description = e.target.value;
        this.setState(() => ({ description })); 
    };

    onNoteChange = e => {
        const note = e.target.value;
        this.setState(() => ({ note }));    
    };

    onAmountChange = e => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        } 
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
        this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calenderFocused: focused }));
    };

    onSubmit = e => {
        e.preventDefault();

        if (!this.state.description ||Â !this.state.amount) {
            this.setState(() => ({ error: 'Please enter a description and a amount'}));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({ 
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                // valueOF() -> momentJS method to get the timestamp back
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };

    render() {
        return (
          <form className="form" onSubmit={this.onSubmit}>
          { this.state.error && <p className="form__error" >{this.state.error}</p>}
              <input type="text"
                  className="input--text"
                  placeholder="Description"
                  autoFocus
                  //ontrolled components for single point of truth
                  value={this.state.description}
                  onChange={this.onDescriptionChange}
              />
              <input type="text" 
                  className="input--text" 
                  placeholder="Amount"
                  value={this.state.amount}
                  onChange={this.onAmountChange}
              />
              <SingleDatePicker 
                  //required
                  date={this.state.createdAt}
                  onDateChange={this.onDateChange}
                  focused={this.state.calenderFocused}
                  onFocusChange={this.onFocusChange}
                  // custom
                  numberOfMonths={1}
                  isOutsideRange={() => false}
              />
              <textarea 
              className="input--textarea"
                  placeholder="Add a Note to your expense"
                  value={this.state.note}
                  onChange={this.onNoteChange}
                  >
              </textarea>
              <div>
                <button className="button">Save Expense</button>
              </div>
              
          </form>
        )
    }
}