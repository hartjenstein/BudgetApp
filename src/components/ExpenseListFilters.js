import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calenderFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calenderFocused) => {  
        this.setState(() => ({ calenderFocused }));
    }

    onTextFilterChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortByChange = (e) => {
        e.target.value === 'date' ?  this.props.sortByDate() : this.props.sortByAmount();
    }

    render() { 
        return (
            <div className="content-container">
              <div className="input-group">
                <div className="input-group__item">
                  <input className="input--text" placeholder="Search Expenses" type="text" value={this.props.filters.text} onChange={this.onTextFilterChange} />
                </div>
                <div className="input-group__item">
                  <select className="input--select"
                  value={this.props.filters.sortBy} // provides controlled input - single point of truth.
                  // Otherwise data would be coming from DOM and could differ
                  onChange={this.onSortByChange}
                  >
                      <option value="date">Date</option>
                      <option value="amount">Amount</option>
                  </select>
                </div>
                <div className="input-group__item">
                    <DateRangePicker 
                    // props required by DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calenderFocused}
                    onFocusChange={this.onFocusChange}
                    // customizitations
                    showClearDates={true}
                    numberOfMonths={1}
                    // enable range beyond current date
                    isOutsideRange={() => false}
                    />
                 </div>
              </div>   
            </div> 
        );
    }
};


// passing in the state object gives us access to the whole state 
// we then return whatever we need from it, in this case all the filters
// we add filters to props
const mapStateToProps = state => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = dispatch => ({
    setTextFilter: text => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});
//connect creates a higher order component that wraps this component 
//with the functionality to connect to the redux store

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);