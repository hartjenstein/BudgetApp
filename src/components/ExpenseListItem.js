import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
 
const ExpenseListItem = ({ id, description, amount, createdAt, note  } ) => (
  <Link className="list__item" to={`/edit/${id}`}>
    <div>
      <h3 className="list__item-title">{description}</h3>
      <span className="list__item-sub-title">{moment(createdAt).format('MMMM Do, YYYY')}</span>
      <span className="list__item-text"> {note &&  <p>Note: {note}</p>}</span>
    </div>
    <h3 className="list__item-data"> {numeral(amount / 100).format('$0,0.00')}</h3>
   
  </Link>
)

export default ExpenseListItem;