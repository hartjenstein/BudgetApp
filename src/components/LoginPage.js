import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';
import CookieBanner from 'react-cookie-banner';

export const LoginPage = ({ startLogin }) => (
    <div className="box-layout">
      <div className="box-layout__box">
        <h1 className="box-layout__title">Budget Control</h1>
        <p>It's time to get your Expenses under control</p>
        <button className="button button--login" onClick={startLogin} >Login with Google</button>
      </div>
      <CookieBanner
      styles={{
        banner: { backgroundColor: 'rgba(60, 60, 60, 0.8)', position: 'fixed', top: 0, left: 0 },
        message: { fontWeight: 400 }
      }}
        message="We use cookies to give you the best online experience.
        Please let us know if you agree."
        onAccept={() => {}}
        cookie="user-has-accepted-cookies" />
    </div>
);


const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);