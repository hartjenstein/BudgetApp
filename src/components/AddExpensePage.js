import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses.js'

// adding export addoitionally to class to make it testable without connect();
export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm 
          onSubmit={this.onSubmit}
          />
        </div>
      
    </div>
    );
  }
}
/// replaced by a class with onSubmit() and mapDispatchToProps() to make it testable 
// - the problem is that the call to dispatch (down below) is referencing 
/* const AddExpensePage = (props) => (
    <div>
      <h1>Add Expense</h1>
      <ExpenseForm 
      onSubmit={(expense) => { */
        // abstracted away (down below) for testing purposes and replaced with a call to onSubmit()
        // props.dispatch(addExpense(expense));
      /*  props.history.push('/');
      }}/>
    </div>
  ); */
///
    //mapDispatchToProps in order to make this component testable with jest
  // we set up the function that connect() takes as a second arguement
  // with mapDispatchProps we pass the call to dispatch and make it testable this way
  //( mapDispatchProps -> a way to return dispatcher functions, abstracting them away from the component itself)
  const mapDispatchToProps = (dispatch) => ({ 
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
  });
    
  
 
  // arguement 
  // if you only need the dispatch method you can call connect without argument
  export default connect(undefined, mapDispatchToProps)(AddExpensePage);