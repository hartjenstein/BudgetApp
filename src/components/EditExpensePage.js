import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';
// Refactor to class based components so we can convert inline functions (meaning in JSX defined) to methods, 
// so that they dont have to be redefined everytime the component gets rendered (for testing purposes )

// there are special props that are passed down by react router like the history object
export class EditExpensePage extends React.Component {
  handleOnSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  handleOnClick= () => {
    this.props.removeExpense({id: this.props.expense.id})
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm 
            expense={this.props.expense}
            onSubmit={this.handleOnSubmit}
        />
        <button 
          onClick={this.handleOnClick}
        >Remove</button>
      </div>
    );
  }
};

// we are searching the state object in the store for the id we get from the props that are passed down by the router 
// we can pass state and props as arguments to the higher order component (hoc) in order to do that
const mapStateToProps = (state, props) => {
  return {
    // we generate a new prop called expense and add it on to props
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch, props) => (
  { editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (data) => dispatch(removeExpense(data))
  });

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);