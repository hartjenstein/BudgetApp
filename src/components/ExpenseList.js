import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

//exporting the unconnected function for testing with enzyme
export const ExpenseList = (props) => (
    <div className="content-container">
      <div className="list__header">
        <div className="show-for-mobile">Expenses</div>
        <div className="show-for-desktop">Expense</div>
        <div className="show-for-desktop">Amount</div>
      </div>
        <div className="list__body">
          {
              props.expenses.length === 0 ? (
                <div className="list__item list__item--message">
                  <span>No expenses</span>
                </div>
      
              ) : props.expenses.map((expense) => (
                  <ExpenseListItem 
                  key={expense.id}
                  {...expense}
                  />
          ))}
        </div>
    </div>
);

//Connecting the component to the redux store with a higher order component (hoc)
// Connect returns another function which 
// creates a higher order component in order to connect to the store
// the stores state gets passed in and we return the props we need out of the state

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

export default connect(mapStateToProps)(ExpenseList);
