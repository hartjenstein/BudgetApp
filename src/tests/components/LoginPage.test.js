import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../components/LoginPage';

let startLogin, wrapper;

beforeEach(() => {
  startLogin = jest.fn();
  wrapper = shallow(<LoginPage startLogin={startLogin}/>);   
});

test('should render LoginPage', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should call startLogout on button click', () => {
  wrapper.find('button').simulate('click'),
  expect(startLogin).toHaveBeenCalled();
});