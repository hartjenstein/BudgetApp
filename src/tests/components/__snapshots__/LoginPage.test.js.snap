// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render LoginPage 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout__title"
    >
      Budget Control
    </h1>
    <p>
      It's time to get your Expenses under control
    </p>
    <button
      className="button button--login"
      onClick={[Function]}
    >
      Login with Google
    </button>
  </div>
  <CookieBanner
    buttonMessage="Got it"
    cookie="user-has-accepted-cookies"
    cookieExpiration={
      Object {
        "years": 1,
      }
    }
    cookies={
      Cookies {
        "HAS_DOCUMENT_COOKIE": true,
        "changeListeners": Array [],
        "cookies": Object {},
        "hooks": undefined,
      }
    }
    dismissOnScroll={true}
    dismissOnScrollThreshold={0}
    message="We use cookies to give you the best online experience. Please let us know if you agree."
    onAccept={[Function]}
    styles={
      Object {
        "banner": Object {
          "backgroundColor": "rgba(60, 60, 60, 0.8)",
          "left": 0,
          "position": "fixed",
          "top": 0,
        },
        "message": Object {
          "fontWeight": 400,
        },
      }
    }
  />
</div>
`;
